### PR TITLE
flowlogs: new vpc flow log format

### PR DESCRIPTION
**Pull request by KalmanMeth**
This PR contains code to implement vpc flow log support via the flow exporter.
@hunchback @eranra For your review.

### CODE REVIEW

**Comment by KalmanMeth** [comment]
run skydive-scale-tests

File: contrib/exporters/vpclogs/mod/object_header.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		augmentedObject[key] = value
+	}
+	augmentedObject["version"] = version
+	augmentedObject["capture_start_time"] = startTime
+	augmentedObject["capture_end_time"] = endTime
+	augmentedObject["number_of_flow_logs"] = len(flows)
+	augmentedObject["flow_logs"] = flows
+	// TBD can we do better with state?
+	augmentedObject["state"] = "skip data"
+
```
Comment (jim-chou) [inline contrib/exporters/vpclogs/mod/object_header.go]: `collector_crn`, `attached_endpoint_type`, and `vpc_crn` are non-optional fields that are not present here
`network_interface_id` and `instance_crn` are optional fields that are not present

Will these be added later, or will we request these be dropped?

File: contrib/exporters/vpclogs/mod/transform.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+type VpclogsFlow struct {
+	FinishType          flow.FlowFinishType `json:"-"`
+	Flow                *flow.Flow          `json:"-"`
+	LastUpdateMetric    *flow.FlowMetric    `json:"-"`
+	Metric              *flow.FlowMetric    `json:"-"`
+	Status              string              `json:"-"`
+	Start               int64               `json:"-"`
+	Protocol            string              `json:"-"`
+	TID                 string              `json:"-"`
+	StartTime           int64               `json:"start_time,omitempty"`
```
Comment (jim-chou) [inline contrib/exporters/vpclogs/mod/transform.go]: These times are still defined as int64s, but the spec was updated to say that output should be RFC3389 in UTC format. Is there code in the json serialization to do this formatting?

Same comment applies to the times in the header object

**Comment by hunchback** [comment]
@KalmanMeth - please note that all activity on exporters had been moved to repo skydive-fllow-exporter - the exporter code here is going to soon be removed. i am closing this PR please reopen in the correct repo. That said i will provide feedback on the code here.

File: contrib/exporters/vpclogs/mod/transform.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+const (
+	VpcActionReject string = "R"
+	VpcActionAccept string = "A"
+)
+
+// VpclogsFlow represents a vpc log flow
+// this structure will be updated when the fields are finalized
+type VpclogsFlow struct {
+	FinishType          flow.FlowFinishType `json:"-"`
```
Comment (hunchback) [inline contrib/exporters/vpclogs/mod/transform.go]: why do these fields appear here if not serialized? please remove all fields annotated with hyphen.

File: contrib/exporters/vpclogs/mod/transform.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	LastUpdateMetric    *flow.FlowMetric    `json:"-"`
+	Metric              *flow.FlowMetric    `json:"-"`
+	Status              string              `json:"-"`
+	Start               int64               `json:"-"`
+	Protocol            string              `json:"-"`
+	TID                 string              `json:"-"`
+	StartTime           int64               `json:"start_time,omitempty"`
+	EndTime             int64               `json:"end_time,omitempty"`
+	ConnectionStartTime int64               `json:"connection_start_time,omitempty"`
+	// TBD - compute direction
```
Comment (hunchback) [inline contrib/exporters/vpclogs/mod/transform.go]: what do you mean by compute direction?

File: contrib/exporters/vpclogs/mod/transform.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	v.CumulativePacketsFromInitiator = v.Metric.ABPackets
+	v.CumulativeBytesFromTarget = v.Metric.BABytes
+	v.CumulativePacketsFromTarget = v.Metric.BAPackets
+	v.BytesFromInitiator = v.LastUpdateMetric.ABBytes
+	v.PacketsFromInitiator = v.LastUpdateMetric.ABPackets
+	v.BytesFromTarget = v.LastUpdateMetric.BABytes
+	v.PacketsFromTarget = v.LastUpdateMetric.BAPackets
+
+	// We don't usually see the report a finished flow
+	// TBM These 2 fields need to be handled differently
```
Comment (hunchback) [inline contrib/exporters/vpclogs/mod/transform.go]: s/TBM/TBD/

File: contrib/exporters/vpclogs/mod/transform.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	} else if v.Protocol == "ICMP" {
+		v.TransportProtocol = 1
+	}
+
+	// TBD fill in direction
+}
+
+// Transform transforms a flow before being stored
+func (ft *vpclogsFlowTransformer) Transform(f *flow.Flow) interface{} {
+	logging.GetLogger().Debugf("flow = %s", f)
```
Comment (hunchback) [inline contrib/exporters/vpclogs/mod/transform.go]: please remove debug messages left from development

File: contrib/exporters/vpclogs/mod/transform.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	updateCount := ft.setUpdateCount(f)
+	status := ft.getStatus(f, updateCount)
+	logging.GetLogger().Debugf("Transform, UUID = %s, status = %s, updateCount = %d", f.UUID, status, updateCount)
```
Comment (hunchback) [inline contrib/exporters/vpclogs/mod/transform.go]: same here

File: contrib/exporters/vpclogs/vpclogs.yml.default [hunk 8]
```diff
@@ -17,8 +17,17 @@ pipeline:
       # - internal
       # - other
   transform:
-    sa:
+    type: vpclogs
+    secadvisor:
       # exclude_started_flows: true
+  objheader:
```
Comment (hunchback) [inline contrib/exporters/vpclogs/vpclogs.yml.default]: these should be part of transform step ... not seperate

File: contrib/exporters/core/object_header.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
+/*
```
Comment (hunchback) [inline contrib/exporters/core/object_header.go]: remove file ... step should be part of transform

File: contrib/exporters/vpclogs/main.go [hunk 10]
```diff
@@ -19,13 +19,14 @@ package main
 
 import (
 	"github.com/skydive-project/skydive/contrib/exporters/core"
-	secadvisor "github.com/skydive-project/skydive/contrib/exporters/secadvisor/mod"
+	"github.com/skydive-project/skydive/contrib/exporters/vpclogs/mod"
```
Comment (hunchback) [inline contrib/exporters/vpclogs/main.go]: use alias 'vpclogs' making it clearer

File: contrib/exporters/vpclogs/mod/object_header.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	return augmentedObject
+}
+
+func NewObjHeaderVpc(cfg *viper.Viper) (interface{}, error) {
+	keyValues := cfg.GetStringMapString("pipeline.objheader.vpclogs")
+	logging.GetLogger().Infof("Defined object header fields:")
+	for key, value := range keyValues {
+		logging.GetLogger().Infof("          %s: %s", key, value)
```
Comment (hunchback) [inline contrib/exporters/vpclogs/mod/object_header.go]: remove log

**Comment by hunchback** [comment]
@KalmanMeth -please explain what is the purpose of object header

File: contrib/exporters/vpclogs/mod/object_header.go (hunk 1 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/object_header.go]: These various IDs are not known to Skydive, so they must be provided by the higher-level management plane. These should be provided in the vpclogs.yml file and are incorporated into the flow logs via the keyValues in the objectHeaderVpc structure. They are then copied directly in to the header of the flow object.

File: contrib/exporters/vpclogs/mod/transform.go (hunk 3 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/transform.go]: Some fields are stored to save intermediate information. These fields are later used to derive the fields that are in the output specification. Some fields may be used when computing the action Accet/Reject, and they need to be available at that time. They are kept as internal fields and not included in the final output.

File: contrib/exporters/vpclogs/mod/transform.go (hunk 4 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/transform.go]: The next field in the structure (according to the spec) is `direction`. The code to compute the direction is not yet written, so this comment is a place-holder to remind us that we still need to implement this.

File: contrib/exporters/vpclogs/mod/transform.go (hunk 5 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/transform.go]: OK

File: contrib/exporters/vpclogs/mod/transform.go (hunk 6 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/transform.go]: If logging level is not set to DEBUG, then these messages won't show up in the output. These messages (when enabled using the DEBUG flag) allow us to demonstrate to team-members (and to ourselves) that things are running as designed. There are numerous such Debug logging statements throughout the Skydive code.

File: contrib/exporters/vpclogs/vpclogs.yml.default (hunk 8 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/vpclogs.yml.default]: The object header fields can't be part of the transform step with the way the exporter code is currently organized. The transform operates on a single flow log entry. The header fields are generated and placed in the pipeline after we have all the flows in hand and  just before the encoding and compressing.  This is done at a certain point in the function flushFlowsToObject,

File: contrib/exporters/vpclogs/mod/object_header.go (hunk 11 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/object_header.go]: This log is important to inform the user of which IDs were defined in the yml file and will be included in the object headers.

**Comment by KalmanMeth** [comment]
The spec of the vpc flow logs states that the format of the object must look like the following:
{
    "version": "0.0.1",
    "attached_endpoint_type": "vnic",
    "capture_end_time": "2008-09-15T15:53:00Z",
    "capture_start_time": "2008-09-15T15:00:00Z",
    "network_interface_id": "crn",
    "collector_crn": "crn",
    "vsi_crn": "crn",
    "vpc_crn": "crn"
    "state": "ok",
    "flow_logs": [
        {
            "action": "accepted",
            "start_time": "2008-09-15T15:53:00.000000Z",
            "end_time": "2008-09-15T15:40:00.000000Z",
            "connection_start_time": "2008-09-15T15:00:00Z",
            "direction": "outbound",
.........

Therefore we have code to produce the object_header according to spec.

File: contrib/exporters/vpclogs/mod/transform.go (hunk 2 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/transform.go]: OK. I will update this.

File: contrib/exporters/vpclogs/mod/transform.go (hunk 4 above)
Comment (hunchback) [reply contrib/exporters/vpclogs/mod/transform.go]: you may consider to add the field but leave it empty (and then if empty don't serialize)

File: contrib/exporters/vpclogs/mod/transform.go (hunk 6 above)
Comment (hunchback) [reply contrib/exporters/vpclogs/mod/transform.go]: I suggest you minimize of such message only leaving the really essential ones

File: contrib/exporters/vpclogs/vpclogs.yml.default (hunk 8 above)
Comment (hunchback) [reply contrib/exporters/vpclogs/vpclogs.yml.default]: then its' a mangle step

Comment (KalmanMeth) [reply contrib/exporters/vpclogs/vpclogs.yml.default]: Then we will need multiple mangle steps; one for the object headers and then one to compute the action Accept/Reject. OK

File: contrib/exporters/vpclogs/main.go (hunk 10 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/main.go]: OK

File: contrib/exporters/vpclogs/mod/transform.go (hunk 4 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/mod/transform.go]: Yes, this is what is currently done.

File: contrib/exporters/vpclogs/vpclogs.yml.default (hunk 8 above)
Comment (KalmanMeth) [reply contrib/exporters/vpclogs/vpclogs.yml.default]: Mangle step is not sufficient. We need to report the number of flow logs being stored in the object. After mangle, we go into store_buffered, which then regulates the number of objects according to various parameters. So we need to intercede after we know the number of flows actually being written to the object.

Comment (KalmanMeth) [reply contrib/exporters/vpclogs/vpclogs.yml.default]: Perhaps we need to make a new store_vpc object that parallels store_buffered.

**Pull request closed without merging**